### PR TITLE
[Team 4] Update agent

### PR DESCRIPTION
**Pull request by hussain2603**
# Summary

Please ignore Training Agent since this is only used for our training purposes.
The Final Agent is an evolutionary agent. This is the only agent that will run in the simulations.

## Additional Information

Some final agent configurations are subject to change. The implementation is final. 

## Test Plan

Tested with our own agent and all other agents currently in main.

### CODE REVIEW

File: pkg/agents/team4/finalAgent/agent.go [hunk 1]
```diff
@@ -36,153 +49,202 @@ type CustomAgentEvo struct {
...
 
-// Data struct to hold evo param values
-type LoadedData struct {
-	FoodToEat  []int
-	DaysToWait []int
-}
+/*------------------------AGENT INITIALISATION------------------------*/
 
-func InitaliseParams(baseAgent *infra.Base) CustomAgentEvoParams {
+func initaliseParams(baseAgent *infra.Base) CustomAgentEvoParams {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:55]:
```suggestion
func initialiseParams(baseAgent *infra.Base) CustomAgentEvoParams {
```

Comment (jzzheng22) [reply pkg/agents/team4/finalAgent/agent.go:55]: ideally fix but whatever

File: pkg/agents/team4/finalAgent/agent.go [hunk 2]
```diff
@@ -36,153 +49,202 @@ type CustomAgentEvo struct {
...
+		trustCoefficients:        []float64{2, 4, 8},
+		lastFoodTaken:            0,
+		intendedFoodToTake:       0,
+		agentResponseMemory:      make(map[uuid.UUID][]int),
+		messageCounter:           0,
+		msgToSendBuffer:          []messages.Message{},
+		sentMessages:             []messages.Message{},
+		responseMessages:         []messages.Message{},
+		requestLeaveFoodMessages: []messages.Message{},
+		otherMessageBuffer:       []messages.Message{},
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:86]: Better to use `make([]messages.Messages, 0)` for these I think

File: pkg/agents/team4/finalAgent/agent.go [hunk 3]
```diff
@@ -36,153 +49,202 @@ type CustomAgentEvo struct {
...
+		timesBlanked := len(a.params.sentMessages) - a.params.receivedMessagesCount
+		a.addToGlobalTrust(-a.params.trustCoefficients[0] * float64(timesBlanked))
+
+		//reset the arrays and maps when we have reshuffled
+		a.params.agentResponseMemory = make(map[uuid.UUID][]int)
+		a.params.msgToSendBuffer = []messages.Message{}
+		a.params.sentMessages = []messages.Message{}
+		a.params.responseMessages = []messages.Message{}
+		a.params.requestLeaveFoodMessages = []messages.Message{}
+		a.params.otherMessageBuffer = []messages.Message{}
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:173]: Ideally use `make` I think

File: pkg/agents/team4/finalAgent/agent.go [hunk 4]
```diff
@@ -36,153 +49,202 @@ type CustomAgentEvo struct {
...
+				log.Error("Simulation - team4/agentTraining.go: \t AlreadyEatenError: Have you already eaten?")
 			default:
 			}
 		}
 	}
 
-	a.Log("team4EvoAgent reporting status:", infra.Fields{"floor": a.Floor(), "hp": a.HP(), "FoodToEat": calculatedAmountToEat, "DaysToWait": a.params.daysToWait, "foodEaten": foodEaten})
+	//log agent state
+	a.Log("team4EvoAgent reporting status:", infra.Fields{"floor": a.Floor(), "hp": a.HP(), "FoodToEat": a.params.intendedFoodToTake, "WaitProbability": a.params.waitProbability, "foodEaten": foodEaten})
 }
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:250]: Please break this up into functions for readability:

- You can put everything inside the `if a.hasReshuffled()` statement into a function
- The second if statement should be in a function too

File: pkg/agents/team4/finalAgent/agent.go [hunk 5]
```diff
@@ -36,153 +49,202 @@ type CustomAgentEvo struct {
...
+	var foodEaten food.FoodType
 	var err error
-	calculatedAmountToEat := food.FoodType(0)
 
-	if (a.Age()-a.params.ageLastEaten) >= a.params.daysToWait[a.params.currentPersonality][a.params.healthStatus] || a.params.healthStatus == 0 {
-		a.params.locked = false
-		calculatedAmountToEat = food.FoodType(a.params.traumaScaleFactor * float64(a.params.foodToEat[a.params.currentPersonality][a.params.healthStatus]))
-		foodEaten, err = a.TakeFood(food.FoodType(calculatedAmountToEat))
-		a.params.ageLastEaten = a.Age()
+	if rand.Intn(100) >= (a.params.waitProbability[a.params.currentPersonality][a.params.healthStatus]-a.params.craving) && !a.HasEaten() && a.PlatformOnFloor() {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:213]: I have no idea what this is comparing - you should probably put the `waitProbability` thing into a separate variable

File: pkg/agents/team4/finalAgent/messages.go [hunk 6]
```diff
@@ -89,62 +85,70 @@ func (a *CustomAgentEvo) HandleAskFoodTaken(msg messages.AskFoodTakenMessage) {
...
 }
 
 func (a *CustomAgentEvo) HandleRequestLeaveFood(msg messages.RequestLeaveFoodMessage) {
-	response := true
-	if a.params.globalTrust < a.params.globalTrustLimit {
-		response = false
+	amount := msg.Request()
+	response := false
+	if a.CurrPlatFood()-a.params.intendedFoodToTake >= food.FoodType(amount) {
+		response = true
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/messages.go:97]:
```suggestion
	response := a.CurrPlatFood()-a.params.intendedFoodToTake >= food.FoodType(amount)
```

File: pkg/agents/team4/finalAgent/treaty.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+// CONDITION OF FLOOR
+// Reject all treaties with the conditionop of FLOOR having > or >= because we don't want to be in a position of loss in any case
+
+// Reject all treaties with duration > max days in critical
+// If request is leave food amount or leave percent food and we are asked to leave 100% of the food then reject cause we would need to eat.
+
+func (a *CustomAgentEvo) HandleProposeTreaty(msg messages.ProposeTreatyMessage) {
+	treaty := msg.Treaty()
+
+	// Switch Case for rejecting treaties
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:31]:
```suggestion
```

File: pkg/agents/team4/finalAgent/treaty.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+	switch {
+	case len(a.ActiveTreaties()) > 0: // If one treaty already signed then we reject any other incoming ones
+		fallthrough
+	case treaty.Request() == messages.Inform: // No idea what inform actually does/how to use it
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:35]: Prefer moving these comments to above each `case` so it's more readable. Likewise for the rest of the switch statement, but ultimately up to you

File: pkg/agents/team4/finalAgent/treaty.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+		toFloor = a.Floor() - 1
+	}
+	propogateTreaty := messages.NewProposalMessage(a.ID(), a.Floor(), toFloor, treaty)
+	a.SendMessage(propogateTreaty)
+}
+
+/*------------------------HANDLING OTHER AGENTS REPSONSES TO OUR TREATY PROPOSALS------------------------*/
+
+func (a *CustomAgentEvo) HandleTreatyResponse(msg messages.TreatyResponseMessage) {
+	if msg.Response() { // Check if there is a response
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:99]: The comment is wrong - `msg.Response()` indicates whether or not they signed the treaty?

File: pkg/agents/team4/finalAgent/treaty.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+		_, ok := a.ActiveTreaties()[msg.TreatyID()] // Check their treaty is valid in our memory
+		if ok {
+			treaty := a.ActiveTreaties()[msg.TreatyID()] // Get the treaty from memory
+			treaty.SetCount(treaty.SignatureCount() + 1) // Update the signature count to our accepted respondent
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:103]: You could just call `SignTreaty()`

File: pkg/agents/team4/finalAgent/treaty.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+		} else if treaty.RequestOp() == messages.GE && a.CurrPlatFood()-a.params.intendedFoodToTake <= food.FoodType(treaty.RequestValue()) {
+			a.params.intendedFoodToTake = food.FoodType(math.Max(0, float64(a.CurrPlatFood()-food.FoodType(treaty.RequestValue()))))
+
+		} else if treaty.RequestOp() == messages.LT && a.CurrPlatFood()-a.params.intendedFoodToTake > food.FoodType(treaty.RequestValue()) {
+			a.params.intendedFoodToTake = food.FoodType(math.Max(0, float64(a.CurrPlatFood()-food.FoodType(treaty.RequestValue())+1)))
+
+		} else if treaty.RequestOp() == messages.LE && a.CurrPlatFood()-a.params.intendedFoodToTake >= food.FoodType(treaty.RequestValue()) {
+			a.params.intendedFoodToTake = food.FoodType(math.Max(0, float64(a.CurrPlatFood()-food.FoodType(treaty.RequestValue()))))
+		}
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:191]: Ideally put this in a separate function and call it inside the case

File: pkg/agents/team4/finalAgent/treaty.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+		} else if treaty.RequestOp() == messages.GE && a.CurrPlatFood()-a.params.intendedFoodToTake <= a.CurrPlatFood()*food.FoodType(treaty.RequestValue())/100 {
+			a.params.intendedFoodToTake = food.FoodType((int(a.CurrPlatFood()) * (1 - treaty.RequestValue()/100)))
+
+		} else if treaty.RequestOp() == messages.LT && a.CurrPlatFood()-a.params.intendedFoodToTake > a.CurrPlatFood()*food.FoodType(treaty.RequestValue())/100 {
+			a.params.intendedFoodToTake = food.FoodType((int(a.CurrPlatFood())*(1-treaty.RequestValue()/100) + 1))
+
+		} else if treaty.RequestOp() == messages.LE && a.CurrPlatFood()-a.params.intendedFoodToTake >= a.CurrPlatFood()*food.FoodType(treaty.RequestValue())/100 {
+			a.params.intendedFoodToTake = food.FoodType((int(a.CurrPlatFood()) * (1 - treaty.RequestValue()/100)))
+		}
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:208]: Ideally put this in a separate function and call it inside the case

File: pkg/agents/team4/finalAgent/treaty.go [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+		} else if treaty.RequestOp() == messages.GT && a.CurrPlatFood()-a.params.intendedFoodToTake < food.FoodType(treaty.RequestValue()) {
+			a.params.intendedFoodToTake = food.FoodType(math.Max(0, float64(a.CurrPlatFood()-food.FoodType(treaty.RequestValue())-1)))
+
+		} else if treaty.RequestOp() == messages.GE && a.CurrPlatFood()-a.params.intendedFoodToTake <= food.FoodType(treaty.RequestValue()) {
+			a.params.intendedFoodToTake = food.FoodType(math.Max(0, float64(a.CurrPlatFood()-food.FoodType(treaty.RequestValue()))))
+
+		} else if treaty.RequestOp() == messages.LT && a.CurrPlatFood()-a.params.intendedFoodToTake > food.FoodType(treaty.RequestValue()) {
+			a.params.intendedFoodToTake = food.FoodType(math.Max(0, float64(a.CurrPlatFood()-food.FoodType(treaty.RequestValue())+1)))
+
+		} else if treaty.RequestOp() == messages.LE && a.CurrPlatFood()-a.params.intendedFoodToTake >= food.FoodType(treaty.RequestValue()) {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:189]: Convert to case and combine with GE condition?

File: pkg/agents/team4/finalAgent/treaty.go [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+		} else if treaty.RequestOp() == messages.GT && a.CurrPlatFood()-a.params.intendedFoodToTake < a.CurrPlatFood()*food.FoodType(treaty.RequestValue())/100 {
+			a.params.intendedFoodToTake = food.FoodType((int(a.CurrPlatFood())*(1-treaty.RequestValue()/100) - 1))
+
+		} else if treaty.RequestOp() == messages.GE && a.CurrPlatFood()-a.params.intendedFoodToTake <= a.CurrPlatFood()*food.FoodType(treaty.RequestValue())/100 {
+			a.params.intendedFoodToTake = food.FoodType((int(a.CurrPlatFood()) * (1 - treaty.RequestValue()/100)))
+
+		} else if treaty.RequestOp() == messages.LT && a.CurrPlatFood()-a.params.intendedFoodToTake > a.CurrPlatFood()*food.FoodType(treaty.RequestValue())/100 {
+			a.params.intendedFoodToTake = food.FoodType((int(a.CurrPlatFood())*(1-treaty.RequestValue()/100) + 1))
+
+		} else if treaty.RequestOp() == messages.LE && a.CurrPlatFood()-a.params.intendedFoodToTake >= a.CurrPlatFood()*food.FoodType(treaty.RequestValue())/100 {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:206]: Convert to case and combine with GE condition?

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go [hunk 15]
```diff
@@ -27,86 +54,98 @@ func (a *CustomAgentEvo) typeAssertRequestMessage(requestMessage messages.Messag
...
 		return
 	}
-	if food.FoodType(reqMsg.Request()) >= a.NeighbourFoodEaten() {
-		a.Log("Team4: For Requested Food to Take greater then or equal neighbour food eaten", infra.Fields{"Request_amt": reqMsg.Request(), "Food_on_our_level": a.NeighbourFoodEaten(), "global_trust": a.params.globalTrust})
-		a.AddToGlobalTrust(a.params.coefficients[1])
+	if food.FoodType(reqMsg.Request()) >= a.neighbourFoodEaten() {
+		a.Log("Team4: For Requested Food to Take greater then or equal neighbour food eaten", infra.Fields{"Request_amt": reqMsg.Request(), "Food_on_our_level": a.neighbourFoodEaten(), "global_trust": a.params.globalTrust})
+		a.addToGlobalTrust(a.params.trustCoefficients[1])
 		return
 	} else {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:87]: Don't need to put this else here

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go [hunk 16]
```diff
@@ -27,86 +54,98 @@ func (a *CustomAgentEvo) typeAssertRequestMessage(requestMessage messages.Messag
...
-func (a *CustomAgentEvo) CheckForResponse(msg messages.BoolResponseMessage) {
+/*------------------------HANDLING TRUST ON RESPONSES ------------------------*/
+
+func (a *CustomAgentEvo) verifyResponses() {
 	if len(a.params.responseMessages) > 0 {
 		for i, respMsg := range a.params.responseMessages { // Iterate through each response message
 			resMsg, err := a.typeAssertResponseMessage(respMsg)
 			if err != nil {
 				log.Error(err)
 			} else {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:101]: Put a return and get rid of the else

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go [hunk 17]
```diff
@@ -27,86 +54,98 @@ func (a *CustomAgentEvo) typeAssertRequestMessage(requestMessage messages.Messag
...
+					a.updateGlobalTrustReqTakeFood(resMsg, sentMsg)
+					isHandled = true
+				}
+				if isHandled {
+					removeMatchingSentMessage(a, resMsg)
+					a.params.responseMessages = remove(a.params.responseMessages, i)
 				}
 			}
 		}
 		return
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:117]: Why do you need this return?

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go [hunk 18]
```diff
@@ -27,86 +54,98 @@ func (a *CustomAgentEvo) typeAssertRequestMessage(requestMessage messages.Messag
...
-						} else if a.params.lastFoodTaken+a.CurrPlatFood() != a.params.lastPlatFood && sentMsg.MessageType() == messages.RequestTakeFood && a.Floor()-msg.SenderFloor() == -1 {
-							a.updateGlobalTrustReqTakeFood(msg, sentMsg)
-						}
-						a.params.sentMessages = remove(a.params.sentMessages, j) // Remove the accessed response/sent messages from memory
-						a.params.responseMessages = remove(a.params.responseMessages, i)
-						break
-					}
+				sentMsg := getMatchingSentMessage(a, resMsg)
+				isHandled := false
+				if a.PlatformOnFloor() && sentMsg.MessageType() == messages.RequestLeaveFood && a.Floor()-resMsg.SenderFloor() == 1 { // Check if there are any responses messages.
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:104]: Move the comment above the if for readability

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go [hunk 19]
```diff
@@ -27,86 +54,98 @@ func (a *CustomAgentEvo) typeAssertRequestMessage(requestMessage messages.Messag
...
+					a.updateGlobalTrustReqLeaveFood(resMsg, sentMsg)
+					isHandled = true
+				} else if a.params.lastFoodTaken+a.CurrPlatFood() != a.params.lastPlatFood && sentMsg.MessageType() == messages.RequestTakeFood && a.Floor()-resMsg.SenderFloor() == -1 {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:107]: Can you write the condition across two lines for readability?

File: pkg/agents/team4/finalAgent/utils.go [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+func getMatchingSentMessage(a *CustomAgentEvo, resMsg messages.ResponseMessage) messages.Message {
+	var retMsg messages.Message
+	for _, sentMsg := range a.params.sentMessages { // Iterate through each sent message
+		if resMsg.RequestID() == sentMsg.ID() { // matches request to relevant sentID
+			retMsg = sentMsg
+			break
+		}
+	}
+	return retMsg
+}
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/utils.go:39]:
```suggestion
	for _, sentMsg := range a.params.sentMessages { // Iterate through each sent message
		if resMsg.RequestID() == sentMsg.ID() { // matches request to relevant sentID
			return sentMsg
		}
	}
}
```

File: pkg/agents/team4/finalAgent/utils.go [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+func removeMatchingSentMessage(a *CustomAgentEvo, resMsg messages.ResponseMessage) {
+	for i, sentMsg := range a.params.sentMessages { // Iterate through each sent message
+		if resMsg.RequestID() == sentMsg.ID() { // matches request to relevant sentID
+			a.params.sentMessages = remove(a.params.sentMessages, i)
+			break
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/utils.go:45]:
```suggestion
			return
```
More clear to use return over break I think (but check it still works)

File: pkg/agents/team4/finalAgent/utils.go [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+	for _, sentMsg := range a.params.sentMessages { // Iterate through each sent message
+		if resMsg.RequestID() == sentMsg.ID() { // matches request to relevant sentID
+			retMsg = sentMsg
+			break
+		}
+	}
+	return retMsg
+}
+
+func removeMatchingSentMessage(a *CustomAgentEvo, resMsg messages.ResponseMessage) {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/utils.go:41]: Should probably be a member function?

File: pkg/agents/team4/finalAgent/utils.go [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+	} else if currentHp <= healthInfo.WeakLevel+2*healthLevelSeparation { //normal
+		return 2
+	} else { //strong
+		return 3
+	}
+}
+
+/*------------------------UTILITIES FOR MATCHING SENT MESSAGES------------------------*/
+
+func getMatchingSentMessage(a *CustomAgentEvo, resMsg messages.ResponseMessage) messages.Message {
```
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/utils.go:30]: Should probably be a member function?

**Review by jzzheng22 — CHANGES_REQUESTED** [review]
ty for submitting pr early bbygrills much appreciated <3

File: pkg/agents/team4/trainingAgent/generateNewBestAgents.py [hunk 24]
```diff
@@ -2,220 +2,188 @@
...
+    return agent
+
+def generate_mixed_agent(agent_list, number_of_agents_to_use, attribute): # We define a mixed agent from parameters of the previous best agents
+
+    picked_best_agents = agent_list[:number_of_agents_to_use]
+
+    bands = []
+    for i in range(number_of_agents_to_use): # determines probability of each agent within the number of agents to use
+        bands.append(((1/number_of_agents_to_use) * (1/(i+1))))
+
```
Comment (woonmoon) [inline pkg/agents/team4/trainingAgent/generateNewBestAgents.py:48]: MINOR: would be more readable in list comprehension `bands = [ (1/number_of_agents_to_use) * (1/(i+1))) for i in range(number_of_agents_to_use) ]`

Comment (woonmoon) [reply pkg/agents/team4/trainingAgent/generateNewBestAgents.py:48]: IMO a lot of the for-loops in this file could be condensed for readability but that's also minor

File: pkg/agents/team4/trainingAgent/getLifeExpectancy.py [hunk 25]
```diff
@@ -17,58 +17,62 @@
 for i in range(len(lines)):
     logs.append(json.loads(lines[i]))
```
Comment (woonmoon) [inline pkg/agents/team4/trainingAgent/getLifeExpectancy.py:19]: MINOR: `logs = [ json.loads(lines[i]) for i in range(len(lines)) ]` - just a suggestion non-necessary.

File: pkg/agents/team4/finalAgent/utils.go [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+/*------------------------REMOVE A PARTICULAR MESSAGE FROM OUR MESSAGE ARRAY------------------------*/
+
+func remove(slice []messages.Message, s int) []messages.Message {
+	return append(slice[:s], slice[s+1:]...)
+}
+
+/*------------------------RETURNS HEALTH STATUS INFORMATION------------------------*/
+
```
Comment (woonmoon) [inline pkg/agents/team4/finalAgent/utils.go:15]: I think you can remove the `else if`

File: pkg/agents/team4/finalAgent/utils.go (hunk 20 above)
Comment (hussain2603) [reply pkg/agents/team4/finalAgent/utils.go:39]: it needs to have a return at the end for it to compile, that's why we've done it like this

**Review by woonmoon — COMMENTED** [review]
LGTM, clearly following honest treaties - made some minor suggestions but feel free to dismiss. <3

File: pkg/agents/team4/finalAgent/treaty.go (hunk 13 above)
Comment (hussain2603) [reply pkg/agents/team4/finalAgent/treaty.go:189]: they have different conditions in the rest of the statement

File: pkg/agents/team4/finalAgent/utils.go (hunk 22 above)
Comment (hussain2603) [reply pkg/agents/team4/finalAgent/utils.go:41]: I think it didn't make sense for us to have this functions be a member function of the agent

File: pkg/agents/team4/trainingAgent/getLifeExpectancy.py (hunk 25 above)
Comment (hussain2603) [reply pkg/agents/team4/trainingAgent/getLifeExpectancy.py:19]: the training agent was just for us and won't be used in the final simulation :)

File: pkg/agents/team4/finalAgent/utils.go (hunk 20 above)
Comment (jzzheng22) [reply pkg/agents/team4/finalAgent/utils.go:41]: Ah right, I think if you have a `return nil` at the end then it would work.

**Review by jzzheng22 — APPROVED** [review]
I think it's fine, I've left one suggestion but it's fine if you don't want.

Ideally wait for another approval if possible

**Review by SaraFFdez — APPROVED** [review]
LGTM

File: pkg/agents/team4/finalAgent/utils.go (hunk 20 above)
Comment (JaafarRammal) [reply pkg/agents/team4/finalAgent/utils.go:41]: our yeah, retMsg is undefined if the loop doesn't match anything, is that okay with your agent behaviour?

**Review by JaafarRammal — APPROVED** [review]
Only that little nil retMessage comment otherwise GTG!

**Pull request merged**